### DISCUSSION TITLE
Base plate issues with Lego Sensor

### DISCUSSION THREAD

**Discussion by TekyTeky** [Novakasa/brickrail#127] [General]
I know you were going to have some downtime but I need some help about an issue that others may also get.

I moved the Brickrail Trains and Switches to my BAP layout.
I have an outer ring route that would be easy to convert.
I added the Red and Blue Block sensors and the same switch hubs. But I'm having a lot of issues with some of the train color sensors and the green baseplate. Not all of the layout has it but all of the outer loop does.

At first it seems to be one area, so I added Redish Brown in-between the track to cover the green. 
Most of the trains seemed to be ok at first. But one train just doesn't go far before it stops with the Marker error like I had previously.
It is always after a Block, any block, even with Priors.

On my test layout that I setup last week, without any green underneath, the same train works fine.

So I guess I need to exclude Green from the chroma, even though all the block sensors are Blue and Red.
I also noticed you added more colours to select for the block sensors.

I can do more tests tomorrow, by adding a green baseplate to the original Brickrail layout or even changing the color sensor as I have more.... unless you have an idea.

See attached of typical areas where it stops, but it also stops on a Lego Switch where there is a square matt area on it.

This is the matt area it always stops just after.. although it could be the green before it.
![IMG_9272](https://user-images.githubusercontent.com/87248517/236662734-862c424a-c1a3-4730-9025-f1e47b670df7.jpeg)

This is typical of an area it stops right in the middle. 
![IMG_9273](https://user-images.githubusercontent.com/87248517/236662777-673ea477-2fc8-4ffd-8209-753d719f885f.jpeg)

![IMG_9274](https://user-images.githubusercontent.com/87248517/236662794-f6e4e7c0-d238-441e-9de0-5ca61711cff3.jpeg)

**Comment by TekyTeky** [comment]
UPDATE- I changed the Color Sensor but it's still the same.. batteries are also ok.

**Comment by Novakasa** [comment]
OK yeah, this issue doesn't surprise me with the way the detection is currently implemented. There is definitely things I can do, but I need to think about it for a bit. I think the easiest would be to add some "ignore" colors or hues. Unfortunately this will be quite hard for me to test currently, so I guess I'll rely on you again :P

I am pretty sure the matte switch should not be the issue, it's probably the green around it.

**Reply by TekyTeky** [reply]
Yeah whatever you want tested let me know. No rush, take it easy and have a break. It does take a bit to set up a new layout on existing track, even though all the hubs worked previously. Initially I had some BLE crashes. I think as I ran BAP and Brickrail on the same pc at the same time. But after a few updates to the FW and restarts it is ok now. I updated to Alpha3 , I'm not sure if it was different from the last master I had but I ended up selecting Download for all the hubs anyway.

**Reply by Novakasa** [reply]
I just pushed a commit that tries to solve your issue.

Now, if any sensor has a color different than "None", the train will only react to colors that are actually set in the virtual layout. This means, if you set any color to _not_ "None", make sure that every color that you use in your physical layout is actually set in the virtual layout at least once. Otherwise, the train will simply ignore that color.

This should hopefully solve your issue, if you don't use any green markers, and you have all your marker colors set in the virtual layout at least once. You can test this with the current master build.

**Reply by TekyTeky** [reply]
Thanks, will try once I can get the new version to connect without the BLE error, as reported separately.

**Comment by TekyTeky** [comment]
This is at the same time as the BLE error I logged tonight.
It is still getting Red and Yellow mixed up, randomly. 
2 consecutive times it got to this yellow prior and then stopped at X.
After the first time I changed chroma to 4000 but still it got confused.
Previously this block hasn't caused any issues.
Same layout.brl as the other error.
This train had not gone in reverse in this session.

OK, I reduced the speed to 40, it's a new train, new sensor and fresh batteries, so maybe it was going to fast and that seems to have resolved the issue... weird.
![image](https://github.com/Novakasa/brickrail/assets/87248517/721fc24a-f4f7-4679-bb80-e625f19d7299)
[Crash_Logs-Color.zip](https://github.com/Novakasa/brickrail/files/11485496/Crash_Logs-Color.zip)
